File: shoring/flywheel/src/main/java/com/panopset/flywheel/CommandReplace.java
package com.panopset.flywheel;

import java.io.StringWriter;
import com.panopset.compat.Stringop;

/**
 * <h1>p - Replace</h1>
 *
 * <p>
 * In this example, &quot;Joe&quot; will be replaced with &quot;Mary&quot;.
 * </p>
 * 
 * <pre>
 *
 * ${&#064;p name}Joe${&#064;q}${&#064;r name}Mary${&#064;q}
 *
 * </pre>
 * 
 * <p>
 * In this example, a prior replacement is undone.
 * </p>
 * 
 * <pre>
 *
 * ${&#064;r name}${name}${&#064;q}
 *
 * </pre>
 *
 */
public class CommandReplace extends MatchableCommand implements
    UserMatchableCommand {

  /**
   * Short HTML text for publishing command format in an HTML document.
   *
   * @return <b>${&#064;r name}</b>.
   */
  public static String getShortHtmlText() {
    return "${&#064;r name}";
  }

  public CommandReplace(final TemplateLine templateLine, final String innerPiece,
      final Template template) {
    super(templateLine, innerPiece, template);
  }

  @Override
  public final void resolve(final StringWriter notUsed) {
    StringWriter sw = new StringWriter();
    resolveMatchedCommands(sw);
    String toReplace = getTemplate().getFlywheel().get(getParams());
    String key = getParams();
    if (Stringop.isPopulated(toReplace)) {
      key = toReplace;
    }
    getTemplate().getFlywheel().getReplacements()
        .add(new String[] { key, sw.toString() });
  }
}

File: code/shoring/flywheel/src/main/kotlin/com/panopset/flywheel/CommandReplace.kt
package com.panopset.flywheel

import com.panopset.compat.Stringop
import java.io.StringWriter

/**
 * <h1>p - Replace</h1>
 *
 *
 *
 * In this example, &quot;Joe&quot; will be replaced with &quot;Mary&quot;.
 *
 *
 * <pre>
 *
 * ${&#064;p name}Joe${&#064;q}${&#064;r name}Mary${&#064;q}
 *
</pre> *
 *
 *
 *
 * In this example, a prior replacement is undone.
 *
 *
 * <pre>
 *
 * ${&#064;r name}${name}${&#064;q}
 *
</pre> *
 *
 */
class CommandReplace(
    templateLine: TemplateLine, innerPiece: String,
    template: Template
) : MatchableCommand( templateLine, innerPiece, template), UserMatchableCommand {
    override fun resolve(sw: StringWriter) {
        val tsw = StringWriter()
        resolveMatchedCommands(tsw)
        val toReplace = template.flywheel.getEntry(getParams())
        var key = getParams()
        if (Stringop.isPopulated(toReplace)) {
            key = toReplace
        }
        template.flywheel.replacements
            .add(arrayOf(key, tsw.toString()))
    }

    companion object {
        val shortHtmlText: String
            /**
             * Short HTML text for publishing command format in an HTML document.
             *
             * @return **${&#064;r name}**.
             */
            get() = "\${&#064;r name}"
    }
}
